src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Ec2Metadata.scala

```diff
@@ -0,0 +1,152 @@
+   * @return some context or None in case of any error including 3 sec timeout
+   */
+  def getInstanceContextBlocking: Option[SelfDescribingJson] =
+    try {
+      Some(Await.result(getInstanceContextFuture, 3 seconds))
+    }
+    catch {
+      case _: Throwable => None
+    }
+
+  /**
+   * Tries to GET self-describing JSON with instance identity
+   *
+   * @return future JSON with identity data
+   */
+  def getInstanceContextFuture: Future[SelfDescribingJson] =
+    getInstanceIdentity.map(SelfDescribingJson(instanceIdentitySchema, _))
+
+  /**
+   * Tries to GET instance identity document for EC2 instance
+   * 
+   * @return future JSON object with identity data
+   */
+  def getInstanceIdentity: Future[JObject] = {
+    val instanceIdentityDocument = pipeline(Get(instanceIdentityUri))
+    instanceIdentityDocument.map(_.entity.asString).map { (resp: String) =>
+      parseOpt(resp) match {
+        case Some(jsonObject: JObject) if jsonObject.values.keys.exists(_ == "instanceId") =>
+          jsonObject
+        case _ =>
+          throw new Exception("Document can not be parsed")
+      }
+    }
+  }
+
+  /**
+   * Recursively parse AWS EC2 instance metadata to get whole metadata
+   *
+   * @param url full url to the endpoint (usually http://169.254.169.254/latest/meta-data/)
+   * @return future JSON object with metadata
+   */
+  def getMetadata(url: String): Future[JObject] = {
+    val key = url.split("/").last
+    if (!url.endsWith("/")) { // Leaf
+      getContent(url).map { value => key -> JString(value) }
+    } else {                  // Node
+      val sublinks = getContents(url)
+      val subnodes: Future[List[JObject]] = sublinks.flatMap { links =>
+        Future.sequence { links.map { link => getMetadata(url + link) } }
+      }
+      val mergedObject = subnodes.map { _.fold(JObject(Nil))(_.merge(_)) }
+      mergedObject.map(key -> _)
+    }
+  }
+
+  // URL regex to for `transformUrl`
+  private val publicKey = ".*/latest/meta-data/public-keys/(\\d+)\\=[A-Za-z0-9-_]+$".r
+
+  /**
+   * Handle URL which should be handled in different ways
+   * e.g. we can't GET public-keys/0-key-name, we should change it to public-keys/0
+   * to get data
+   *
+   * @param url current URL
+   * @return modified URL if we're trying to get on of special cases
+   */
+  def transformUrl(url: String): String = url match {
+    case publicKey(i) => (url.split("/").dropRight(1) :+ i).mkString("/") + "/"
+    case _            => url
+  }
+
+  /**
+   * Get URL content (for leaf-link)
+   *
+   * @param url leaf URL (without slash at the end)
+   * @return future value
+   */
+  private def getContent(url: String): Future[String] =
+    pipeline(Get(url)).map(_.entity.asString)
+
+  /**
+   * Get content of node-link
+   *
+   * @param url node url (with slash at the end)
+   * @return future list of sublinks
+   */
+  private def getContents(url: String): Future[List[String]] =
+    getContent(url).map(_.split('\n').toList)
+}
```

Comment:
Actually that will make a lot of sense and clean API without blocking call/missing context. Should I do it now?